File: array/maximum_index.py
# User function Template for python3
class Solution:
    def maxIndexDiff(self, arr, n):
        # code here
        left = n * [0]
        right = n * [0]
        left[0] = arr[0]
        for i in range(1, n):
            left[i] = min(left[i - 1], arr[i])

        right[n - 1] = arr[n - 1]
        for i in range(n - 2, -1, -1):
            right[i] = max(right[i + 1], arr[i])

        i, j, ans = 0, 0, 0
        while i < n and j < n:
            if left[i] <= right[j]:
                ans = max(ans, j - i)
                j += 1
            else:
                i += 1

        return ans


# {
# Driver Code Starts
if __name__ == "__main__":
    t = int(input())
    while t > 0:
        num = int(input())
        arr = [int(x) for x in input().strip().split()]
        ob = Solution()
        print(ob.maxIndexDiff(arr, num))
        t -= 1
# } Driver Code Ends

File: array/maximum_index.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;

// } Driver Code Ends
class Solution {
public:
  // A[]: input array
  // N: size of array
  // Function to find the maximum index difference.
  int maxIndexDiff(int A[], int N) {
    // Your code here
    vector<int> dp(N);
    dp[0] = 0;
    for (int i = 1; i < N; ++i) {
      dp[i] = i;
      if (A[i] >= A[i - 1]) {
        for (int j = 0; j <= dp[i - 1]; ++j) {
          if (A[i] >= A[j]) {
            dp[i] = j;
            break;
          }
        }
      } else {
        for (int j = dp[i - 1]; j < i; ++j) {
          if (A[i] >= A[j]) {
            dp[i] = j;
            break;
          }
        }
      }
    }
    int ans = 0;
    for (int i = 0; i < N; ++i) {
      ans = max(i - dp[i], ans);
    }

    return ans;
  }
};

//{ Driver Code Starts.

/* Driver program to test above functions */
int main() {
  int T;
  // testcases
  cin >> T;
  while (T--) {
    int num;
    // size of array
    cin >> num;
    int arr[num];

    // inserting elements
    for (int i = 0; i < num; i++)
      cin >> arr[i];
    Solution ob;

    // calling maxIndexDiff() function
    cout << ob.maxIndexDiff(arr, num) << endl;
  }
  return 0;
}
// } Driver Code Ends
